csrc/host_ir/pass/deallocate.cpp

```diff
@@ -0,0 +1,52 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <host_ir/pass/deallocate.h>
+
+namespace nvfuser::hir {
+
+void insertDeallocations(HostIrContainer* hic) {
+  const std::vector<Expr*>& topLevelExprs = hic->topLevelExprs();
+  std::unordered_map<TensorView*, std::vector<nvfuser::Expr*>::size_type>
+      lastUse;
+  for (std::vector<nvfuser::Expr*>::size_type i = 0; i < topLevelExprs.size();
+       i++) {
+    Expr* expr = topLevelExprs[i];
+    if (expr->isA<Deallocate>()) {
+      auto* deallocate = expr->as<Deallocate>();
+      auto* tv = deallocate->allocation();
+      NVF_ERROR(
+          lastUse.count(tv) > 0, "Tried to deallocate unknown TensorView");
+      lastUse.erase(tv);
+    } else {
+      for (auto* val : expr->inputs()) {
+        if (!val->isA<TensorView>()) {
+          continue;
+        }
+        auto tv = val->as<TensorView>();
+        lastUse[tv] = i;
+      }
+    }
+  }
+
+  std::map<std::vector<nvfuser::Expr*>::size_type, std::vector<TensorView*>>
```

Comment:
```suggestion
  std::map<int64_t, std::vector<TensorView*>>
```

Reply:
Also, you don't need a map for this. Simply convert last_use to a vector of `pair<int64_t, TensorView*>` and sort by the expression index. 

Later, you can `for (auto&& [i, tv] : the_vector | std::views::reverse`.

Reply:
Isn't this the same thing as using std::map and reverse iterating through it though?

Reply:
Also, in order to account for expressions being used multiple times, wouldn't you have to search the vector for each insertion?

Reply:
Oh, you meant replace rmap with last_use_by_index, not replace last_use. That makes sense, thanks for the explanation.